### PR TITLE
resize label matrix

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

File: src/execution_plan/ops/shared/update_functions.h [hunk 1]
```diff
@@ -13,15 +13,15 @@
...
 } PendingUpdateCtx;
 
 // commit all updates described in the array of pending updates
 void CommitUpdates
 (
 	GraphContext *gc,
 	dict *updates,
+	RG_Matrix add_labels,
+	RG_Matrix remove_labels,
 	EntityType type
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.h:25]: All arguments should be commented.

File: src/execution_plan/ops/shared/update_functions.h [hunk 2]
```diff
@@ -33,6 +33,8 @@ void EvalEntityUpdates
 	GraphContext *gc,
 	dict *node_updates,
 	dict *edge_updates,
+	RG_Matrix *add_labels,
+	RG_Matrix *remove_labels,
 	const Record r,
 	const EntityUpdateEvalCtx *ctx,
 	bool allow_null
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.h:40]: All arguments should be commented.

File: src/execution_plan/ops/shared/update_functions.h [hunk 3]
```diff
@@ -33,6 +33,8 @@ void EvalEntityUpdates
 	GraphContext *gc,
 	dict *node_updates,
 	dict *edge_updates,
+	RG_Matrix *add_labels,
+	RG_Matrix *remove_labels,
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.h:37]: I'm not convinced we need `RG_Matrix` maybe we can use a `HYPER-SPARSE GrB_Matrix`

File: src/execution_plan/ops/shared/update_functions.c [hunk 4]
```diff
@@ -57,28 +59,21 @@ void CommitUpdates
 
 	dictEntry *entry;
 	dictIterator *it = HashTableGetIterator(updates);
-	MATRIX_POLICY policy = Graph_GetMatrixPolicy(gc->g);
-	Graph_SetMatrixPolicy(gc->g, SYNC_POLICY_NOP);
+
+	if(type == ENTITY_NODE) {
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.c:63]: add comment: `// updating nodes labels`

File: src/execution_plan/ops/shared/update_functions.c [hunk 5]
```diff
@@ -57,28 +59,21 @@ void CommitUpdates
...
+		UpdateLabels(gc, add_labels, remove_labels, true);
+	}
 
 	while((entry = HashTableNext(it)) != NULL) {
 		PendingUpdateCtx *update = HashTableGetVal(entry);
 
-		// if entity has been deleted, perform no updates
-		if(GraphEntity_IsDeleted(update->ge)) continue;
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.c:67]: Why was this line removed?

File: src/execution_plan/ops/shared/update_functions.c [hunk 6]
```diff
@@ -180,25 +176,44 @@ void EvalEntityUpdates
...
 	} else {
 		// update context already exists
 		update = (PendingUpdateCtx *)HashTableGetVal(entry);
 	}
 
-	if(array_len(ctx->add_labels) > 0 && update->add_labels == NULL) {
-		update->add_labels = array_new(const char *, array_len(ctx->add_labels));
+	if(array_len(ctx->add_labels) > 0 && *add_labels == NULL) {
+		RG_Matrix_new(add_labels, GrB_BOOL, Graph_RequiredMatrixDim(gc->g),
+			Graph_RequiredMatrixDim(gc->g));
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.c:188]: the dimensions should be: number of nodes in the graph X number of labels in the graph

File: src/execution_plan/ops/shared/update_functions.c [hunk 7]
```diff
@@ -180,25 +176,44 @@ void EvalEntityUpdates
...
 	}
 	
-	array_union(update->add_labels, ctx->add_labels, strcmp);
-	array_union(update->remove_labels, ctx->remove_labels, strcmp);
+	for (uint i = 0; i < array_len(ctx->add_labels); i++) {
+		const char *label = ctx->add_labels[i];
+		const Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
+		if(s == NULL) {
+			s = AddSchema(gc, label, SCHEMA_NODE, true);
+		}
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.c:201]: The `AddSchema` operation MUST be performed under write lock as it might reallocate graph's labels array while READ queries has access to.

File: src/execution_plan/ops/shared/update_functions.c [hunk 8]
```diff
@@ -180,25 +176,44 @@ void EvalEntityUpdates
...
+		int schema_id = Schema_GetID(s);
+		RG_Matrix_setElement_BOOL(*add_labels, ENTITY_GET_ID(entity), schema_id);
+	}
+
+	for (uint i = 0; i < array_len(ctx->remove_labels); i++) {
+		const char *label = ctx->remove_labels[i];
+		const Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
+		if(s == NULL) {
+			s = AddSchema(gc, label, SCHEMA_NODE, true);
+		}
```
Comment (swilly22) [inline src/execution_plan/ops/shared/update_functions.c:212]: Do NOT create a scheme in this case, consider
`MATCH (n:N) REMOVE n:NoneExistingSchema`
There's no harm "removing" a non existing schema from a node. obviously the node isn't labeled under such schema...

Simply `continue`

File: src/execution_plan/ops/op_update.c [hunk 9]
```diff
@@ -167,4 +168,12 @@ static void UpdateFree(OpBase *ctx) {
...
 
 	raxStop(&op->it);
+
+	if(op->add_labels) {
+		RG_Matrix_free(&op->add_labels);
+	}
+
+	if(op->remove_labels) {
+		RG_Matrix_free(&op->remove_labels);
+	}
```
Comment (swilly22) [inline src/execution_plan/ops/op_update.c:178]: I believe this logic needs to be also performed in `UpdateReset`
Generally speaking I see there's some diffs between `UpdateFree` and `UpdateReset` take this opportunity to review `UpdateReset` and make sure we haven't missed anything.

File: src/execution_plan/ops/op_merge.h [hunk 10]
```diff
@@ -16,20 +16,22 @@
...
+	Record *input_records;                   // Records generated by the bound variable stream
+	Record *output_records;                  // Records to be emitted by this operation
+	rax *on_match;                           // Updates to be performed on a successful match
+	rax *on_create;                          // Updates to be performed on creation
+	raxIterator on_match_it;                 // Iterator for traversing ON MATCH update contexts
+	raxIterator on_create_it;                // Iterator for traversing ON CREATE update contexts
+	dict *node_pending_updates;              // Pending updates to apply, generated
+	dict *edge_pending_updates;              // Pending updates to apply, generated
+	RG_Matrix add_labels;                    // Labels to add to nodes
+	RG_Matrix remove_labels;                 // Labels to remove from nodes
```
Comment (swilly22) [inline src/execution_plan/ops/op_merge.h:34]: if you've already modified this, please make sure the spacing is correct, 2 char from the longest line.

File: src/execution_plan/ops/op_merge.c [hunk 11]
```diff
@@ -129,8 +131,6 @@ OpBase *NewMergeOp
 
 	op->on_match             = on_match;
 	op->on_create            = on_create;
-	op->node_pending_updates = NULL;
-	op->edge_pending_updates = NULL;
 	
 	// set our Op operations
 	OpBase_Init((OpBase *)op, OPType_MERGE, "Merge", MergeInit, MergeConsume,
```
Comment (swilly22) [inline src/execution_plan/ops/op_merge.c:136]: How come `MERGE` doesn't have a `reset` function ?

File: src/graph/graph_hub.c [hunk 12]
```diff
@@ -428,15 +428,133 @@ void UpdateNodeLabels
...
 	}
 }
 
+void UpdateLabels
+(
+	GraphContext *gc,            // graph context to update the entity
+	RG_Matrix add_labels,        // labels to add to nodes
+	RG_Matrix remove_labels,     // labels to remove from nodes
+	bool log                     // log this operation in undo-log
+) {
```
Comment (swilly22) [inline src/graph/graph_hub.c:446]: comments should start 2 characters from the longest line

File: src/graph/graph_hub.c [hunk 13]
```diff
@@ -428,15 +428,133 @@ void UpdateNodeLabels
...
+				UndoLog_RemoveLabels(undo_log, ENTITY_GET_ID(node),
+						remove_labels_ids, remove_labels_index);
+				EffectsBuffer_AddRemoveLabelsEffect(eb, ENTITY_GET_ID(node),
+						remove_labels_ids, remove_labels_index);
 			}
 		}
 	}
 }
 
+void UpdateLabels
```
Comment (swilly22) [inline src/graph/graph_hub.c:440]: missing documentation comment

File: src/graph/graph_hub.c [hunk 14]
```diff
@@ -428,15 +428,133 @@ void UpdateNodeLabels
...
+	int *labels_ids = array_new(int, 0);
+	GrB_Index prev_node_id = 0;
+	GrB_Index node_id;
+	GrB_Index label_id;
+	if(add_labels != NULL) {
+		RG_MatrixTupleIter it = {};
+		RG_MatrixTupleIter_attach(&it, add_labels);
+
+		while(RG_MatrixTupleIter_next_BOOL(&it, &node_id, &label_id, NULL) == GrB_SUCCESS) {
+			if(node_id != prev_node_id && array_len(labels_ids) > 0) {
```
Comment (swilly22) [inline src/graph/graph_hub.c:464]: Add comment
`// commit node's labels`

File: src/graph/graph_hub.c [hunk 15]
```diff
@@ -428,15 +428,133 @@ void UpdateNodeLabels
...
+
+			if(!node_labeled) {
+				Schema *s = GraphContext_GetSchemaByID(gc, (int)label_id, SCHEMA_NODE);
+				// append label id
+				array_append(labels_ids, label_id);
+				// add to index
+				Node n;
+				Graph_GetNode(gc->g, node_id, &n);
+				Schema_AddNodeToIndices(s, &n);
+			}
```
Comment (swilly22) [inline src/graph/graph_hub.c:486]: I think this is a bit too much,
maybe we should perform `add_labels = add_labels - node_labels` as a preliminary step, this will make sure `add_labels` contains only "new" labels. removing the need for checking for each node if it is already labeled under `label_id`.

File: src/graph/graph_hub.c [hunk 16]
```diff
@@ -428,15 +428,133 @@ void UpdateNodeLabels
...
+			prev_node_id = node_id;
+		}
+
+		if(array_len(labels_ids) > 0) {
```
Comment (swilly22) [inline src/graph/graph_hub.c:490]: comment: `// labelling last node`

**Review by swilly22 — CHANGES_REQUESTED** [review]
Overall interesting idea,
This could be a killer improvement IF the operation of updating node labels was common

File: src/util/arr.h [hunk 17]
```diff
@@ -325,29 +325,6 @@ __extension__({                                         \
         }                                       \
     })                                          \
 
-// dest = dest ∪ src
-// treats both src and dest as sets
-// [1,2,3] ∪ [2,4,4] = [1,2,3,4,4]
-// TODO: if either src or dest are large and we're allowed to rearrange
-// elements position then consider sorting
-#define array_union(dest, src, cmp)                  \
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/util/arr.h:333]: no

File: src/undo_log/undo_log.c [hunk 18]
```diff
@@ -163,11 +163,13 @@ static void _UndoLog_Rollback_Set_Labels
...
-		Graph_RemoveNodeLabels(g, update_labels_op->node.id,
+		Graph_RemoveNodeLabels(g, update_labels_op->node_id,
 				update_labels_op->label_ids, labels_count);
 
-		_index_delete_node_with_labels(ctx, &(update_labels_op->node),
-				update_labels_op->label_ids, labels_count);
+		Node n;
+		Graph_GetNode(g, update_labels_op->node_id, &n);
+		_index_delete_node_with_labels(ctx, &n, update_labels_op->label_ids,
+				labels_count);
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/undo_log/undo_log.c:172]: yes because now when we don't have any problem we can use matrix to store the added/removed labels and only if need to undo we need to get the node

File: src/execution_plan/ops/shared/update_functions.c (hunk 5 above)
Comment (AviAvni) [reply src/execution_plan/ops/shared/update_functions.c:67]: no needed anymore now that the delete operation is eager the check in EvalEntityUpdates

File: src/execution_plan/ops/op_merge.c (hunk 11 above)
Comment (AviAvni) [reply src/execution_plan/ops/op_merge.c:136]: add tests

File: src/graph/graph_hub.c (hunk 15 above)
Comment (AviAvni) [reply src/graph/graph_hub.c]: now that we using GrB_Matrix it is harder to do it against RG_Matrix
